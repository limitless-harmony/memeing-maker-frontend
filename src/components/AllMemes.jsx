import React, { Component } from 'react';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { calculateRem } from 'styles';
import { getMany } from 'actions/meme';
import MemeContainer from 'components/MemeContainer';

export class AllMemes extends Component {
  componentDidMount() {
    this.fetchMemes();
  }

  fetchMemes = async () => {
    const { actions, meta } = this.props;
    await actions.getMany(meta.nextPage || 1);
  };

  render() {
    const { memes, meta } = this.props;
    return (
      <StyledFeatured>
        {memes && memes.length ? (
          <>
            <SectionHeading>Available Memes</SectionHeading>
            <MemeContainer
              memes={memes}
              hasNextPage={meta.hasNextPage}
              fetchMore={this.fetchMemes}
            />
          </>
        ) : (
          <SectionHeading>No meme available yet!</SectionHeading>
        )}
      </StyledFeatured>
    );
  }
}

const StyledFeatured = styled.div`
  margin: ${calculateRem(10)} auto ${calculateRem(40)};
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 100%;
  text-align: center;
`;

const SectionHeading = styled.div`
  font-size: ${calculateRem(18)};
  margin: ${calculateRem(16)} auto;
`;

const mapStateToProps = state => ({
  memes: state.meme.memes,
  meta: state.meme.meta,
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators({ getMany }, dispatch),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AllMemes);
