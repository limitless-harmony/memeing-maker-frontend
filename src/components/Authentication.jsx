import React, { Component } from 'react';
import styled from 'styled-components';

import { calculateRem } from 'styles';
import {
  FacebookColored,
  GoogleColored,
  LinkedinColored,
} from 'components/Icons';

class Authentication extends Component {
  handleLogin = strategy => {
    console.log(strategy);
  };

  loginWithGoogle = () => {
    return this.handleLogin('google');
  };

  loginWithFacebook = () => {
    return this.handleLogin('facebook');
  };

  loginWithLinkedIn = () => {
    return this.handleLogin('linkedIn');
  };

  render() {
    return (
      <StyledAuthentication>
        <TextSection>Sign up or log in</TextSection>
        <TextSection>to create, share, & discover</TextSection>
        <TextSection> our full library of meaningful memes.</TextSection>
        <ButtonSection>
          <Button onClick={this.loginWithGoogle}>
            <GoogleColored />
          </Button>
          <Button onClick={this.loginWithFacebook}>
            <FacebookColored />
          </Button>
          <Button onClick={this.loginWithLinkedIn}>
            <LinkedinColored />
          </Button>
        </ButtonSection>
      </StyledAuthentication>
    );
  }
}

const StyledAuthentication = styled.div`
  margin: ${calculateRem(-70)} auto auto;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100%;
  text-align: center;
`;

const TextSection = styled.div`
  align-items: center;
  font-size: ${calculateRem(18)};
`;

const ButtonSection = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  margin: ${calculateRem(20)} auto;
`;

const Button = styled.div`
  margin: 0 ${calculateRem(12)};
`;

export default Authentication;
