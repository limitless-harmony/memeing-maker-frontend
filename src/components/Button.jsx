import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import { calculateRem } from 'styles';
import { white, darken, green } from 'styles/colors';

const Button = ({ color, children, rounded, outline, fullWidth }) => {
  return (
    <StyledButton
      color={color}
      outline={outline}
      rounded={rounded}
      fullWidth={fullWidth}
    >
      {children}
    </StyledButton>
  );
};

export const StyledButton = styled.button`
  border: ${({ outline, color }) => (outline ? `2px solid ${color}` : 'none')};
  background-color: ${({ outline, color }) =>
    outline ? 'transparent' : color};
  color: ${({ outline, color }) => (outline ? color : white)};
  border-radius: ${({ rounded }) =>
    rounded ? calculateRem(4000) : calculateRem(8)};
  text-align: center;
  cursor: pointer;
  letter-spacing: 1px;
  width: ${({ fullWidth }) => (fullWidth ? '100%' : 'auto')};
  font-size: ${calculateRem(18)};
  margin: ${calculateRem(10)};
  padding: ${calculateRem(3)} ${calculateRem(30)};
  &:hover {
    background-color: ${({ color }) => darken(color, 0.05)};
    color: ${white};
  }
`;

Button.propTypes = {
  color: PropTypes.string,
  outline: PropTypes.bool,
  rounded: PropTypes.bool,
  fullWidth: PropTypes.bool,
  children: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
};

Button.defaultProps = {
  color: green,
  outline: false,
  rounded: false,
  fullWidth: false,
  children: 'Submit',
};

export default Button;
