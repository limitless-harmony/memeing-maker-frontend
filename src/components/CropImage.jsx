import React from 'react';
import styled from 'styled-components';
import Cropper from 'react-easy-crop';

const CropImage = ({
  image,
  crop,
  zoom,
  aspect,
  onCropChange,
  onCropComplete,
  onZoomChange,
}) => (
  <StyledCropImage>
    <Cropper
      image={image}
      crop={crop}
      zoom={zoom}
      aspect={aspect}
      onCropChange={onCropChange}
      onCropComplete={onCropComplete}
      onZoomChange={onZoomChange}
    />
  </StyledCropImage>
);

const StyledCropImage = styled.div`
  position: relative;
  width: 100%;
  height: 400px;
`;

export default CropImage;
