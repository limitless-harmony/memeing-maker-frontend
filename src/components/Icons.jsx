import { ReactComponent as Create } from 'assets/icons/create.svg';
import { ReactComponent as ShareButton } from 'assets/icons/share-button.svg';
import { ReactComponent as Facebook } from 'assets/icons/facebook.svg';
import { ReactComponent as Discover } from 'assets/icons/discover.svg';
import { ReactComponent as Ellipsis } from 'assets/icons/ellipsis.svg';
import { ReactComponent as AddImage } from 'assets/icons/add-image.svg';
import { ReactComponent as AddMemo } from 'assets/icons/add-memo.svg';
import { ReactComponent as FacebookColored } from 'assets/icons/facebook-colored.svg';
import { ReactComponent as FeelsReaction } from 'assets/icons/feels-reaction.svg';
import { ReactComponent as GoArrow } from 'assets/icons/go-arrow.svg';
import { ReactComponent as GoogleColored } from 'assets/icons/google-colored.svg';
import { ReactComponent as Instagram } from 'assets/icons/instagram.svg';
import { ReactComponent as LinkedinColored } from 'assets/icons/linkedin-colored.svg';
import { ReactComponent as Linkedin } from 'assets/icons/linkedin.svg';
import { ReactComponent as SocialShare } from 'assets/icons/social-share.svg';
import { ReactComponent as SocialLogins } from 'assets/icons/social-logins.svg';
import { ReactComponent as Submit } from 'assets/icons/submit.svg';
import { ReactComponent as ThanksReaction } from 'assets/icons/thanks-reaction.svg';
import { ReactComponent as ThoughtsReaction } from 'assets/icons/thoughts-reaction.svg';
import { ReactComponent as Twitter } from 'assets/icons/twitter.svg';
import { ReactComponent as WhatsApp } from 'assets/icons/whatsapp.svg';

export {
  Create,
  ShareButton,
  Facebook,
  Discover,
  Ellipsis,
  AddImage,
  AddMemo,
  FacebookColored,
  FeelsReaction,
  GoArrow,
  GoogleColored,
  Instagram,
  LinkedinColored,
  Linkedin,
  SocialShare,
  SocialLogins,
  Submit,
  ThanksReaction,
  ThoughtsReaction,
  Twitter,
  WhatsApp,
};
