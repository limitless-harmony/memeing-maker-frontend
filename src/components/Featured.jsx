import React, { Component } from 'react';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { calculateRem } from 'styles';
import { getFeatured } from 'actions/meme';
import MemeCard from 'components/MemeCard';
import Button from 'components/Button';
import { buttonBorder, black } from 'styles/colors';

export class Featured extends Component {
  async componentDidMount() {
    const { actions } = this.props;
    await actions.getFeatured();
  }

  render() {
    const { memes } = this.props;
    return (
      <StyledFeatured>
        {memes.length ? (
          <>
            <SectionHeading>This Weekâ€™s Featured Memes</SectionHeading>
            <Container>
              {memes.map(meme => (
                // eslint-disable-next-line no-underscore-dangle
                <Meme key={meme._id}>
                  <MemeCard
                    small
                    src={meme.image}
                    topText={meme.topText}
                    bottomText={meme.bottomText}
                  />
                </Meme>
              ))}
            </Container>
            <Button color={buttonBorder} outline rounded>
              <ButtonText to="/login">More</ButtonText>
            </Button>
          </>
        ) : (
          <SectionHeading>No featured meme for this week!</SectionHeading>
        )}
      </StyledFeatured>
    );
  }
}

const StyledFeatured = styled.div`
  margin: ${calculateRem(10)} auto ${calculateRem(40)};
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 100%;
  text-align: center;
`;

const Container = styled.div`
  margin: ${calculateRem(10)} auto ${calculateRem(40)};
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  width: 100%;
  text-align: center;
`;

const Meme = styled.div`
  width: 48.5%;
`;

const SectionHeading = styled.div`
  font-size: ${calculateRem(18)};
  margin: ${calculateRem(16)} auto;
`;

const ButtonText = styled(Link)`
  color: ${black};
  text-decoration: none;
`;

const mapStateToProps = state => ({
  memes: state.meme.featured,
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators({ getFeatured }, dispatch),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Featured);
