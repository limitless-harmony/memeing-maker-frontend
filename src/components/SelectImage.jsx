import React, { Component, Fragment } from 'react';
import styled from 'styled-components';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import { hideModal } from 'actions/modal';
import { selectImage } from 'actions/image';
import { startLoader, stopLoader } from 'actions/loading';
import { calculateRem } from 'styles';
import CropImage from 'components/CropImage';
import Button from 'components/Button';
import ImageIcon from 'components/ImageIcon';
import folder from 'assets/icons/folder.png';
import gallery from 'assets/icons/gallery.png';
import { getCroppedImage, readFile } from 'helpers/image';

export class SelectImage extends Component {
  state = {
    croppedAreaPixels: {},
    imageUrl: '',
    crop: { x: 0, y: 0 },
    zoom: 1,
    aspect: 1 / 1,
  };

  changeText = event => {
    this.setState({
      [event.currentTarget.name]: event.currentTarget.value,
    });
  };

  uploadImageFile = async () => {
    const { imageUrl, croppedAreaPixels } = this.state;
    const { actions } = this.props;
    const img = document.createElement('img');
    img.setAttribute('src', `${imageUrl}`);
    img.onload = () => {
      const base64 = getCroppedImage(img, croppedAreaPixels);
      actions.selectImage(base64);
    };
    actions.hideModal();
  };

  onCropChange = crop => {
    this.setState({ crop });
  };

  onCropComplete = (croppedArea, croppedAreaPixels) => {
    this.setState({ croppedAreaPixels });
  };

  onZoomChange = zoom => {
    this.setState({ zoom });
  };

  onFileChange = async e => {
    const { actions } = this.props;
    if (e.target.files && e.target.files.length > 0) {
      actions.startLoader();
      const imageDataUrl = await readFile(e.target.files[0]);
      actions.stopLoader();
      this.setState({
        imageUrl: imageDataUrl,
      });
    }
  };

  render() {
    const { imageUrl, crop, zoom, aspect } = this.state;
    return (
      <StyledSelectImage>
        <IconSection>
          <Label htmlFor="image">
            <ImageIcon width={50} src={folder} />
            <input
              id="image"
              type="file"
              accept="image/*"
              onChange={this.onFileChange}
            />
          </Label>
          <ImageIcon src={gallery} />
        </IconSection>
        <CropSection>
          {imageUrl && (
            <Fragment>
              <CropImage
                image={imageUrl}
                crop={crop}
                zoom={zoom}
                aspect={aspect}
                style={{ height: '100%' }}
                onCropChange={this.onCropChange}
                onCropComplete={this.onCropComplete}
                onZoomChange={this.onZoomChange}
              />
              <Button outline onClick={this.uploadImageFile}>
                Confirm
              </Button>
            </Fragment>
          )}
        </CropSection>
      </StyledSelectImage>
    );
  }
}

const StyledSelectImage = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  width: 100%;
  box-sizing: border-box;
  text-align: center;
`;

const IconSection = styled.div`
  display: flex;
  width: 100%;
  justify-content: space-evenly;
  align-items: center;
  padding: ${calculateRem(20)} ${calculateRem(40)};
  min-height: ${calculateRem(80)};
  box-sizing: border-box;
`;
const CropSection = styled.div`
  width: 100%;
`;

const Label = styled.label`
  height: 100%;
  input {
    display: none;
  }
`;

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(
    { hideModal, selectImage, startLoader, stopLoader },
    dispatch
  ),
});

export default connect(
  null,
  mapDispatchToProps
)(SelectImage);
