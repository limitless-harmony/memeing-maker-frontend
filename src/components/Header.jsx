import React, { Component } from 'react';
// import PropTypes from 'prop-types';
import styled from 'styled-components';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

import { calculateRem, mobileWidth } from 'styles';
import { ShareButton, Discover, Ellipsis, Create } from 'components/Icons';
import { white } from 'styles/colors';
import { showModal } from 'actions/modal';
import toggleMenu from 'actions/menu';

export class Header extends Component {
  share = () => {
    const { isLoggedIn, actions } = this.props;
    if (!isLoggedIn) {
      return actions.showModal('auth');
    }
    return actions.showModal('share');
  };

  toggleMenu = () => {
    const { actions, menuStatus } = this.props;
    return actions.toggleMenu(!menuStatus);
  };

  render() {
    return (
      <StyledHeader>
        <NavLeft>
          <NavItem first>
            <Link to="/">
              <Discover />
            </Link>
          </NavItem>
          <NavItem>
            <Link to="/create">
              <Create />
            </Link>
          </NavItem>
          <NavItem>
            <ShareButton width="18" height="18" onClick={this.share} />
          </NavItem>
        </NavLeft>
        <NavRight>
          <NavItem last>
            <Ellipsis onClick={this.toggleMenu} />
          </NavItem>
        </NavRight>
      </StyledHeader>
    );
  }
}

export const StyledHeader = styled.header`
  display: flex;
  justify-content: space-between;
  align-items: center;
  height: ${calculateRem(36)};
  padding: 0 ${calculateRem(14)};
  background: ${white};
  position: fixed;
  z-index: 10;
  width: 100%;
  max-width: ${calculateRem(mobileWidth)};
  box-sizing: border-box;
  left: 50%;
  top: 0%;
  transform: translateX(-50%);
`;

const NavItem = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  cursor: pointer;
  margin: ${({ first, last }) =>
    first
      ? `0 ${calculateRem(5)} 0 0`
      : last
      ? `0 0 0 ${calculateRem(5)}`
      : `0 ${calculateRem(5)}`};
`;

const NavSection = styled.div`
  display: flex;
  flex: 1;
`;

const NavLeft = styled(NavSection)`
  justify-content: flex-start;
`;

const NavRight = styled(NavSection)`
  justify-content: flex-end;
`;

Header.propTypes = {
  isLoggedIn: PropTypes.bool.isRequired,
  actions: PropTypes.shape({}).isRequired,
};

const mapStateToProps = state => ({
  menuStatus: state.menu.show,
  isLoggedIn: state.auth.isLoggedIn,
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators({ showModal, toggleMenu }, dispatch),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Header);
