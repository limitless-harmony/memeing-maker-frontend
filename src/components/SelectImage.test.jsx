import React from 'react';
import { shallow } from 'enzyme';

import { SelectImage } from 'components/SelectImage';

import * as imageHelper from 'helpers/image';

const mockFunction = jest.fn();

describe('Testing the SelectImage page component', () => {
  const props = {
    selectedImage: '/images/me.png',
    actions: {
      hideModal: mockFunction,
      selectImage: mockFunction,
    },
  };
  it('renders without crashing', () => {
    const wrapper = shallow(<SelectImage />);
    expect(wrapper.find('div'));
    expect(wrapper).toMatchSnapshot();
  });
  it('invoking changeText updates the state', () => {
    const wrapper = shallow(<SelectImage {...props} />);
    const instance = wrapper.instance();
    expect(wrapper.state('imageUrl')).toBe('');
    const event = {
      currentTarget: {
        name: 'imageUrl',
        value: 'veraclins.com',
      },
    };
    instance.changeText(event);
    expect(wrapper.state('imageUrl')).toBe('veraclins.com');
  });
  it('invoking onFileChange updates the state', async () => {
    const wrapper = shallow(<SelectImage {...props} />);
    const instance = wrapper.instance();
    expect(wrapper.state('imageUrl')).toBe('');
    const blob = new Blob(['image'], { type: 'image/png' });
    blob.lastModifiedDate = '';
    blob.name = 'filename.png';
    const event = {
      target: {
        files: [blob],
      },
    };
    const dataUrl = await imageHelper.readFile(blob);
    await instance.onFileChange(event);
    expect(wrapper.state('imageUrl')).toBe(dataUrl);
  });
  it('invoking onCropChange updates the state', () => {
    const wrapper = shallow(<SelectImage {...props} />);
    const instance = wrapper.instance();
    expect(wrapper.state('crop')).toEqual({ x: 0, y: 0 });
    const crop = { x: 200, y: 200 };
    instance.onCropChange(crop);
    expect(wrapper.state('crop')).toEqual({ x: 200, y: 200 });
  });
  it('invoking onZoomChange updates the state', () => {
    const wrapper = shallow(<SelectImage {...props} />);
    const instance = wrapper.instance();
    expect(wrapper.state('zoom')).toEqual(1);
    const zoom = 2;
    instance.onZoomChange(zoom);
    expect(wrapper.state('zoom')).toEqual(2);
  });
  it('invoking onCropComplete updates the state', () => {
    const wrapper = shallow(<SelectImage {...props} />);
    const instance = wrapper.instance();
    expect(wrapper.state('croppedAreaPixels')).toEqual({});
    const croppedAreaPixels = { x: 200, y: 200 };
    instance.onCropComplete(croppedAreaPixels, croppedAreaPixels);
    expect(wrapper.state('croppedAreaPixels')).toEqual({ x: 200, y: 200 });
  });
  it('invoking uploadImageFile records the cropped image and closes the select-image modal', () => {
    const wrapper = shallow(<SelectImage {...props} />);
    const instance = wrapper.instance();
    instance.uploadImageFile();
    expect(mockFunction).toHaveBeenCalled();
  });
});
