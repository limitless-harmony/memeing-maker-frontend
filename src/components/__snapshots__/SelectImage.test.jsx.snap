// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the SelectImage page component renders without crashing 1`] = `
<styled.div>
  <styled.div>
    <ImageIcon
      src="camera.png"
      width={50}
    />
    <styled.label
      htmlFor="image"
    >
      <ImageIcon
        src="folder.png"
        width={50}
      />
      <input
        accept="image/*"
        id="image"
        onChange={[Function]}
        type="file"
      />
    </styled.label>
    <ImageIcon
      src="gallery.png"
    />
  </styled.div>
  <styled.div />
</styled.div>
`;
